snapcraft/internal/repo.py

```diff
@@ -291,18 +291,24 @@ def get(self, package_names):
         with self._apt.archive(self._cache.base_dir) as apt_cache:
             self._get(apt_cache, package_names)
 
-    def _get(self, apt_cache, package_names):
-        manifest_dep_names = self._manifest_dep_names(apt_cache)
-
+    def _mark_install(self, apt_cache, package_names):
         for name in package_names:
             try:
                 logger.debug(
                     'Marking {!r} (and its dependencies) to be fetched'.format(
                         name))
-                apt_cache[name].mark_install()
+                name_arch, version = _get_pkg_name_parts(name)
+                if version:
+                    _set_pkg_version(apt_cache[name_arch], version)
+                apt_cache[name_arch].mark_install()
             except KeyError:
                 raise PackageNotFoundError(name)
 
+    def _get(self, apt_cache, package_names):
+        manifest_dep_names = self._manifest_dep_names(apt_cache)
+
+        self._mark_install(apt_cache, package_names)
```

Comment:
I think this is going a good direction, but we might as well complete the refactor here. This function is made up of three steps:

1. Mark requested packages for installation
2. Unmark base packages for installation
3. Actually fetch archives.

I say we simply put each in its own function. No real code changes.